src/compiler/tracing.ts

```diff
@@ -0,0 +1,256 @@
+/*@internal*/
+/** Tracing events for the compiler. */
+namespace ts.tracing {
+    let fs: typeof import("fs") | false | undefined;
+
+    let traceCount = 0;
+    let traceFd: number | undefined;
+
+    let legendPath: string | undefined;
+    const legend: TraceRecord[] = [];
+
+    /** Starts tracing for the given project (unless the `fs` module is unavailable). */
+    export function startTracing(configFilePath: string | undefined, traceDir: string, isBuildMode: boolean) {
+        Debug.assert(!traceFd, "Tracing already started");
+
+        if (fs === undefined) {

```

Comment:
In the future we might be able to move some of this logic to a conditionally loaded module, similar to the extended debug information for flow nodes in _~/src/debug/dbg.ts_ loaded by _~/src/compiler/debug.ts_: https://github.com/microsoft/TypeScript/blob/d89635381a65229c5ab5b0c73de15347b86235a2/src/compiler/debug.ts#L655

I've been thinking about moving more of `ts.Debug` into the external file to reduce the file size of _tsc.js_, _typescript.js_, and _tsserver.js_ as well.

This obviously isn't something we need to do now, but we might want to consider putting it on the backlog.